lib/pbench/test/unit/server/test_file_tree.py

```diff
@@ -328,8 +332,11 @@ def mock_rmtree(path: Path, ignore_errors=False):
             assert ignore_errors
             assert path == incoming / f"{tarball_name}.unpack"
 
-        def mock_run(self, command, dir_path, exception, dir_p):
-            raise subprocess.TimeoutExpired("tar", 43)
+        @staticmethod
+        def mock_run(command, dir_path, exception, dir_p):
```

Comment:
Do you really want/need the `@staticmethod` decorator here?  (I don't even know what that does, given that we're in a function scope instead of directly inside a class scope...and I would think it would work fine without it.)

Ditto line 368 and 402.

Reply:
Without `@staticmethod` either it is expecting me to include `self` or throwing error `TestFileTree.test_unpack_tar_subprocess_exception.<locals>.mock_run() takes 4 positional arguments but 5 were given
`

Reply:
Every time I think I understand this stuff, I get proven wrong, again.  Sorry!  😢 

> Note that the transformation from function object to instance method object happens each time the attribute is retrieved from the instance.  [[Ref]](https://docs.python.org/3/reference/datamodel.html#index-35)

> If a class attribute is found that is a user-defined function object, it is transformed into an instance method object whose `__self__` attribute is the instance. Static method and class method objects are also transformed;  [[Ref]](https://docs.python.org/3/reference/datamodel.html#index-49)

> When [a class attribute reference] would yield a static method object, it is transformed into the object wrapped by the static method object.  [[Ref]](https://docs.python.org/3/reference/datamodel.html#index-45)

> Static method objects provide a way of defeating the transformation of function objects to method objects described above. A static method object is a wrapper around any other object, usually a user-defined method object. When a static method object is retrieved from a class or a class instance, the object actually returned is the wrapped object, which is not subject to any further transformation. Static method objects are also callable. Static method objects are created by the built-in [staticmethod()](https://docs.python.org/3/library/functions.html#staticmethod) constructor.  [[Ref]](https://docs.python.org/3/reference/datamodel.html#index-65)

So, I _think_ that when we patch `Tarball.subprocess_run()` and replace it with a simple function reference, we are replacing a static method object with a regular user-defined function object.  Thus, when the CUT makes the call, the function object is transformed into an instance method, and the `self` parameter is added to the beginning of the argument list.

So, in order to make the patch work, we either need to declare the mock with a leading `self` parameter (which is a bit of a misfit, given that the function that we're mocking is a static method), or we need to ensure that the value that we're using to patch the method is a `staticmethod` object and not a simple function object.  One way to achieve that is to use the `@staticmethod` decorator when we declare it.  The other is to construct a `staticmethod` object on the fly:
```
            m.setattr(Tarball, "subprocess_run", staticmethod(mock_run))
```
Given that function decorators really have nothing to do with class definitions per se (that's just where they frequently appear), I guess it's fine to use the `@staticmethod` decorator...it just looks odd to me outside the class definition context.  However, [apparently it is synonymous](https://docs.python.org/3/glossary.html#term-decorator) with using `staticmethod()` explicitly.

So, sorry for the misdirection.  Pick whichever one seems best to you (assuming that they both actually work...).

Reply:
we got to know a little extra 😄

Reply:
Yes, and that's good, I just wish the getting were a little less painful.  🙂